### ISSUE TITLE
Crowdin: Automation of export/import 

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#3601]
Currently the export (of new strings) to Crowdin as well as the import (of approved strings) from Crowdin has to be triggered manually.
This could be changed to either a periodical automated execution or an on demand execution (e.g. as a job on the CI-server).

**Comment by Lineflyer** [comment]
Meanwhile Github offers an UI integration app towards Crowdin:
https://github.com/marketplace/crowdin
That might be interesting to take a look at.

**Comment by rsudev** [comment]
Would be nice, especially if through that option the integration with our branches would work better. But currently I am a bit reluctant to try it in order to not break the working workflow towards the next feature release.

**Comment by Lineflyer** [comment]
My first try was not really succesful. The app enrollment seems to require, that we are not yet using Crowdin at all. It first throws an error, that Crowdin is already authorized to access the repo. Secondly it triggered a new project creation on Crowdin (I have sent a mail to Crowdin to disregard it, as we already have a project running).
Maybe this problem happend as I tried with my personal account but not the organization...anyway I will stop for now as I had the same thought as you had: Do not kill any existing workflows.

Maybe a jenkins job would be easier to implement as the requires scripts are already existing.

**Comment by Lineflyer** [comment]
As I am currently exploring and implementing the Github <-> Crowdin integration for WhereYouGo and got it working right now, how about to use this as well for c:geo?

The current script solution has the big disadvantage, that not everyone can easily use it.
The integration would provide automated PRs from everything approved on Crowdin.

As I am currently dealing with this and we would potentially use the very same user (cgeo-ci-bot) also for the cgeo repository, it would be rather easy to implement this for c:geo as well.

It should not break anything, I would only suggest to drop translating `release` branch and stick to `master` only. 
Reason for this and further details can be found here: https://github.com/cgeo/WhereYouGo/issues/23

Any objections?

**Comment by Lineflyer** [comment]
@rsudev 
Are there any special implementation currently used in the scripts,or do you see any risk points if we would migrate to the crowdin integration via Github app? AFAIR you are the author of the scripts currently used.

**Comment by Lineflyer** [comment]
If no one stops me, I will give it a try maybe tomorrow evening.

**Comment by bekuno** [comment]
We have a special renaming in the crowdin script:
~~~
map_to_crowdin_code () {
    CODE=$1
    [[ "$CODE" == "es" ]] && CODE="es-ES"
    [[ "$CODE" == "sv" ]] && CODE="sv-SE"
    [[ "$CODE" == "pt" ]] && CODE="pt-PT"
    echo "$CODE"
}
~~~

Please do not forget cgeo-contacts.

**Comment by bekuno** [comment]
>     * What would be a good folder to place additional translatable text which is not part of the app (like playstore description, etc.)? In WhereYouGo there was already a `_data` folder for many things which are just to be stored but not part of the app itself. Do we already have something similar already?

We need something similar in the folders cgeo-contacts and main.
Why not create such a folder _data there?

**Comment by bekuno** [comment]
>     * Currently we have our changelog in `main/res/values/changelog_*.xml`. Can we just include these files to be translated with Crowdin (so tranlated versions of these files end up in each value directory) and it will work? I don't know the mechanism behind that file in c:geo.
 
> 
> > ```
> > * What would be a good folder to place additional translatable text which is not part of the app (like playstore description, etc.)? In WhereYouGo there was already a `_data` folder for many things which are just to be stored but not part of the app itself. Do we already have something similar already?
> > ```
> 
> We need something similar in the folders cgeo-contacts and main.
> Why not create such a folder _data there?

Or, we can integrate the playstore etc. description in the res/values/ folder.

**Comment by bekuno** [comment]
>     * Currently we have our changelog in `main/res/values/changelog_*.xml`. 

They are `<string name="changelog_master" translatable="false">` , needs at least to make them translatable.

**Comment by Lineflyer** [comment]
Another question:
I assume the strings for cgeo-calendar on Crowdin are superfluous as it is discontinued?

**Comment by bekuno** [comment]
cgeo-calendar is no project anymore and so superfluous.

**Comment by Lineflyer** [comment]
While checking the diffs I find something strange:
There is a significant percentage showing as untranslated now. Mostly these are strings which do not need to be translated (e.g. platform name like "geocaching.com" or "opencaching.de" or app names like "GPS Locker").
They are not declared as `translatable=false` in our strings, so its only normal they are now in for translation. But why did this not show up before.
If I understand the process right, we should mark them as non-translatable in the strings.xml file, or not? Maybe they have only been manually excluded on Crowdin up to now.

**Comment by Lineflyer** [comment]
To Do list:
- [ ] Some strings without translation need (e.g. own names of apps and platforms, cache sizes (where we decided to use the english name throughout)) have only been blocked locally on Crowdin up to now. These should be marked as non-translatable in the c:geo repository itself. Do they need to be added to `strings_not_translatable.xml` or just individually marked as `translatable=false` within `strings.xml`?
- [ ] French translation (and others?) suddenly contain unicode replacement codes (e.g. instead of the triple point character. This might be a fault caused by needed manual export/import to keep translations on Crowdin not yet imported to the repository
- [ ] Add strings.xml from c-geo contacts into crowdin.yml

**Comment by Lineflyer** [comment]
Converted my ToDo into a project: https://github.com/cgeo/cgeo/projects/13

For here I will only leave the questions where I need help:

- Some strings without translation need (e.g. own names of apps and platforms, cache sizes (where we decided to use the english name throughout)) have only been blocked locally on Crowdin up to now. These should be marked as non-translatable in the c:geo repository itself. Do they need to be added to strings_not_translatable.xml or just individually marked as translatable=false within strings.xml?

**Comment by moving-bits** [comment]
To my understanding the distribution of strings over two different files just serves an easier file comparison. Technically it should be enough to mark them as translatable=false.

File names like `strings.xml` and `strings_not_translatable.xml` are arbitrary (https://developer.android.com/guide/topics/resources/string-resource) and chosen by us. As long as the file(s) reside(s) in the `values` folder we should be on the safe side.

**Comment by Lineflyer** [comment]
> File names like strings.xml and strings_not_translatable.xml are arbitrary (https://developer.android.com/guide/topics/resources/string-resource) and chosen by us. As long as the file(s) reside(s) in the values folder we should be on the safe side.

But seems like in the past we used the convention, that non translatable strings should go to `strings_not_translatable.xml`. While I guess, that Crowdin will follow the `translatable` flag, I think that putting those strings in a dedicated file is better to get an overview.

@moving-bits 
If we want to move strings from one file to another, is there a special sequence we need to keep? AFAIR from the distant past, strings cannot just be placed anywhere in that file...is it right?

**Comment by moving-bits** [comment]
> But seems like in the past we used the convention, that non translatable strings should go to `strings_not_translatable.xml`. While I guess, that Crowdin will follow the `translatable` flag, I think that putting those strings in a dedicated file is better to get an overview.

Yes, it makes it easier to distinguish and to maintain, so I also would like to stay with this convention, although technically it's not necessary.

> If we want to move strings from one file to another, is there a special sequence we need to keep? AFAIR from the distant past, strings cannot just be placed anywhere in that file...is it right?

Not that I know of. AFAICS for plain text strings (in contrast to themes, styles, attributes, menus etc.) you can place the strings anywhere in the files, as long as they stay within the `resources` tag.

**Comment by Lineflyer** [comment]
Does anyone (or the compiler) have a problem if we refrain from using the placeholders (i.e. english versions of strings used in each localized file) until the real translation comes in?
At first I thought it might be nicer to always have complete files, but in practice its hard to review PRs if for each new source strings it bounces back this string to all languages.

From what I read it should anyway fall back to english in the app if a certain resource is not translated, right?

**Comment by moving-bits** [comment]
The compiler doesn't have any problem with it as this is the normal scenario for adding a new string during development - you only add it to the default file in the `values` directory, and either during build or at runtime (I'm not sure about that part) it gets sorted out for the other languages.

**Comment by Lineflyer** [comment]
@moving-bits @bekuno 
The translation data should be consistent now, see PR #8200
I do have one problem left according to the lint report:
https://ci.cgeo.org/job/cgeo%20pull%20request/2641/androidLintResult/new/

Any idea where this comes from? 
I guess in one of the languages a variable is missing, but I do not find where...

**Comment by Lineflyer** [comment]
Nevermind. I found it in the Fillipino strings.

**Comment by Lineflyer** [comment]
First merge has been done.
I will keep this issue open and tonight perform another manual sync (auto sync switched off at the moment) for final check.

For open sub-items (e.g. cgeo-contacts strings, playstore description) I will later use dedicated issues to track them.

@cgeo/team 
Feel free to check current state of `/main/res/values/` on `master` if e.g. the languages codes are correct.
BTW: 
At the moment I realized the same language codes as before. Crowdin however also supports the placeholder `%android_code%` which is intended to be used for Android projects. This will always return language codes like `de-rDE`. If we could constantly use this, there would be no need of manual remappings. Does someone have insights here?

**Comment by Lineflyer** [comment]
New automated integration is done.
Translations should now by synced as described on the updated wiki page:
https://github.com/cgeo/cgeo/wiki/Translation-%28L18N%29

Similar to cgeo/WhereYouGo I decided to only use the `master` branch as new strings on `release` branch are rare. The workflow has been implemented for the main app as well as cgeo-contacts.

AFAICS we did not loose the major part of existing translations. Only some not proofread translations of languages not yet included in c:geo might have been lost on Crowdin.

I will open dedicated issues for some remaining optimizations.
